app/src/main/scala/AppState.scala

```diff
@@ -48,11 +48,12 @@ object AppState:
         case Some(task)                           => GetTaskResult.AcquiredByOther(task)
 
     def updateOrGiveUp(candidates: List[Work.Task]): (AppState, List[Work.Task]) =
-      candidates.foldLeft(state -> Nil) { case ((state, xs), task) =>
-        task.clearAssignedKey match
-          case None                 => (state - task.id, task :: xs)
-          case Some(unAssignedTask) => (state.updated(task.id, unAssignedTask), xs)
-      }
+      candidates.mapAccumulateFilter(state)(_.clearOrGiveUp(_))
+
+    def clearOrGiveUp(task: Work.Task): (AppState, Option[Work.Task]) =
```

Comment:
this refactor is great, and the name as you mentioned can be improved. I think both `updateOrGiveUp` and `clearOrGiveUp` can be renamed to `unassignOrGiveup`.